Modules/xSharePoint/DSCResources/MSFT_xSPSearchRoles/MSFT_xSPSearchRoles.psm1

```diff
@@ -0,0 +1,246 @@
+            $QueryProcessingExists = $true
+        }
+
+        $indexComps = Get-SPEnterpriseSearchComponent -SearchTopology $currentTopology `
+            | Where-Object {($_.GetType().Name -eq "IndexComponent") `
+                -and ($_.IndexPartitionOrdinal -eq $params.Index)}
+
+        $servers = ""
+        foreach ($indexComp in $indexComps) {
+            $servers += $indexComp.ServerName + ","
+        }
+
+        return @{
+            ServiceAppName = $params.ServiceAppName
+            Admin = $AdminExists
+            Crawler = $CrawlerExists
+            ContentProcessing = $ContentProcessingExists
+            AnalyticsProcessing = $AnalyticsProcessingExists
+            QueryProcessing = $QueryProcessingExists
+            InstallAccount = $params.InstallAccount
+            FirstPartitionIndex = $params.FirstPartitionIndex
+            FirstPartitionDirectory = $params.FirstPartitionDirectory
+            FirstPartitionServers = $servers.TrimEnd(",")
+            Ensure = $params.Ensure
+        }
+    }
+    return $result
+}
+
+
+function Set-TargetResource
+{
+    [CmdletBinding()]
+    param
+    (
+        [parameter(Mandatory = $true)]  [System.String]  $ServiceAppName,
+        [parameter(Mandatory = $true)]  [System.Boolean] $Admin,
+        [parameter(Mandatory = $true)]  [System.Boolean] $Crawler,
+        [parameter(Mandatory = $true)]  [System.Boolean] $ContentProcessing,
+        [parameter(Mandatory = $true)]  [System.Boolean] $AnalyticsProcessing,
+        [parameter(Mandatory = $true)]  [System.Boolean] $QueryProcessing,
+        [parameter(Mandatory = $false)] [System.Management.Automation.PSCredential] $InstallAccount,
+        [parameter(Mandatory = $true)]  [ValidateSet("Present","Absent")] [System.String] $Ensure,
+        [parameter(Mandatory = $false)] [System.UInt32]  $FirstPartitionIndex,
+        [parameter(Mandatory = $true)]  [System.String]  $FirstPartitionDirectory,
+        [parameter(Mandatory = $true)]  [System.String]  $FirstPartitionServers
+    )
+
+    $CurrentValues = Get-TargetResource @PSBoundParameters
+
+    Invoke-xSharePointCommand -Credential $InstallAccount -Arguments @($PSBoundParameters, $CurrentValues) -ScriptBlock {
+        $params = $args[0]
+        $CurrentValues = $args[1]
+        $ConfirmPreference = 'None'
+
+        $ssi = Get-SPEnterpriseSearchServiceInstance -Local
+        $ssa = Get-SPEnterpriseSearchServiceApplication
+
+        if($ssi.Status -eq "Offline") {
+            Write-Verbose "Start Search Service Instance"
+            Start-SPEnterpriseSearchServiceInstance -Identity $ssi
+        }
+
+        #Wait for Search Service Instance to come online
+        $online = Get-SPEnterpriseSearchServiceInstance -Identity $ssi; 
+        do {
+            $online = Get-SPEnterpriseSearchServiceInstance -Identity $ssi; 
+            Write-Verbose "Waiting for service: $($online.TypeName)"
+        } 
+        until ($online.Status -eq "Online")
```

Comment:
Search services doesn't usually get stuck, but, I think you need to have a timeout here, don't you?